From f037284faf255937156b78cfe74779b40ff7e123 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Thu, 14 Mar 2024 07:23:36 +0200
Subject: [PATCH] [FSW-402] fs1: build: add linker script

This patch adds linker scripts for different platform
types of fs1. It also adds export text file which will
be used for exporting bootrom symbols to other
components.

Change-Id: I334d009bc074f1c8bdd9fccc528dc28b3c704767
---

diff --git a/arch/fs1/arc/asic_boot_rom.lds b/arch/fs1/arc/asic_boot_rom.lds
new file mode 100644
index 0000000..34786be
--- /dev/null
+++ b/arch/fs1/arc/asic_boot_rom.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ */
+
+INCLUDE boot_rom.lds
diff --git a/arch/fs1/arc/boot_rom.lds b/arch/fs1/arc/boot_rom.lds
new file mode 100644
index 0000000..a8efd5e
--- /dev/null
+++ b/arch/fs1/arc/boot_rom.lds
@@ -0,0 +1,102 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ */
+
+OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
+OUTPUT_ARCH(arc)
+
+MEMORY
+{
+	ICCM	: ORIGIN = 0x00000000, LENGTH = 256K
+	SPI	: ORIGIN = 0x94000000, LENGTH = 16M
+	SBROM	: ORIGIN = 0x78980000, LENGTH = 128K
+	FWSRAM	: ORIGIN = 0x789A0000, LENGTH = 128K
+	DCCM	: ORIGIN = 0x80000000, LENGTH = 128K
+}
+
+REGION_ALIAS("REGION_IVT",	SBROM)
+REGION_ALIAS("REGION_TEXT",	SBROM)
+REGION_ALIAS("REGION_DATA",	DCCM)
+REGION_ALIAS("REGION_SDATA",	DCCM)
+REGION_ALIAS("REGION_SBSS",	DCCM)
+REGION_ALIAS("REGION_BSS",	DCCM)
+REGION_ALIAS("REGION_HEAP",	DCCM)
+REGION_ALIAS("REGION_STACK",	DCCM)
+
+ENTRY(__start)
+
+SECTIONS
+{
+	_dccm_begin = ORIGIN(DCCM);
+	_dccm_end = ORIGIN(DCCM) + LENGTH(DCCM);
+	_persistent_sec_size = 3k;
+	_rom_global_data_size = 512;
+
+	.ivt : {
+		KEEP (*(.ivt));
+	} > REGION_IVT
+
+	.text : {
+		app/crt0_rom.o (.text*)
+		*(.text .text.*)
+		_etext = . ;
+	} > REGION_TEXT
+
+	.data : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
+		_data_after_persist_sec = _etext + _persistent_sec_size;
+		_data = .;
+		/* TODO: Finalize design on persistent and non-persistent section */
+	    KEEP(*(.rom_persistent_section))
+	    . = _data + _rom_global_data_size; /* Reserve 512 memory for persistent data section */
+	    KEEP(*(.rom_manifest_section))
+	    . = _data + _persistent_sec_size; /* Reserve 2560 (3K - 512) memory for persistent data section */
+		_rom_non_persistent_dccm_start = .;
+		*(.data)
+	} > REGION_DATA
+
+	.sdata : AT ( LOADADDR(.data) + SIZEOF(.data) ) {
+		__SDATA_BEGIN__ = .;
+		*(.sdata .sdata.* .gnu.linkonce.s.*)
+		_edata = .;
+		PROVIDE (edata = .);
+	} > REGION_SDATA
+
+	.sbss : {
+		__bss_start = .;
+		PROVIDE (__sbss_start = .);
+		PROVIDE (___sbss_start = .);
+		*(.dynsbss)
+		*(.sbss .sbss.* .gnu.linkonce.sb.*)
+		*(.scommon)
+		PROVIDE (__sbss_end = .);
+		PROVIDE (___sbss_end = .);
+	} > REGION_SBSS
+
+	.bss : {
+		*(.dynbss)
+		*(.bss .bss.* .gnu.linkonce.b.*)
+		*(COMMON)
+		/* Align here to ensure that the .bss section occupies space up to
+		_end. Align after .bss to ensure correct alignment even if the
+		.bss section disappears because there are no input sections. */
+		. = ALIGN(32 / 8);
+		_end = .;
+		PROVIDE (end = .);
+	} > REGION_BSS
+
+	.heap : {
+		. = ALIGN(32 / 8);
+		__start_heap = . ;
+		. = . + (DEFINED(__HEAP_SIZE) ? __HEAP_SIZE : (DEFINED(__DEFAULT_HEAP_SIZE) ? __DEFAULT_HEAP_SIZE : 5k)) ;
+		__end_heap = . ;
+		. = ALIGN(0x8);
+	} > REGION_HEAP
+
+	.stack : {
+		. = ALIGN(32 / 8);
+		__stack = . ;
+		. = . + (DEFINED(__STACK_SIZE) ? __STACK_SIZE : (DEFINED(__DEFAULT_STACK_SIZE) ? __DEFAULT_STACK_SIZE : 8k)) ;
+		__stack_top = . ;
+	} > REGION_STACK
+}
diff --git a/arch/fs1/arc/boot_rom.txt b/arch/fs1/arc/boot_rom.txt
new file mode 100644
index 0000000..81056c6
--- /dev/null
+++ b/arch/fs1/arc/boot_rom.txt
@@ -0,0 +1 @@
+# put list of symbols, one per line, to be exported
diff --git a/arch/fs1/arc/onix_boot_rom.lds b/arch/fs1/arc/onix_boot_rom.lds
new file mode 100644
index 0000000..34786be
--- /dev/null
+++ b/arch/fs1/arc/onix_boot_rom.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ */
+
+INCLUDE boot_rom.lds
diff --git a/arch/fs1/arc/pldm_boot_rom.lds b/arch/fs1/arc/pldm_boot_rom.lds
new file mode 100644
index 0000000..34786be
--- /dev/null
+++ b/arch/fs1/arc/pldm_boot_rom.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ */
+
+INCLUDE boot_rom.lds
